frontend/app/utils/raoidc-utils.server.ts

```diff
@@ -0,0 +1,144 @@
+/**
+ * Utility functions to help with RAOIDC requests.
+ */
+import { getLogger } from './logging.server';
+
+const log = getLogger('oidc-utils.server');
+
+/**
+ * An OIDC authentication server.
+ * Returned by the OIDC server's discovery endpoint.
+ *
+ * @see https://datatracker.ietf.org/doc/html/rfc8414#section-2
+ */
+export interface ServerMetadata extends Record<string, unknown> {
+  authorization_endpoint: string;
+  issuer: string;
+  jwks_uri: string;
+  token_endpoint: string;
+  userinfo_endpoint: string;
+  //
+  // optional fields
+  //
+  end_session_endpoint?: string;
+  introspection_endpoint_auth_methods_supported?: Array<string>;
+  introspection_endpoint_auth_signing_alg_values_supported?: Array<string>;
+  mtls_endpoint_aliases?: {
+    device_authorization_endpoint?: string;
+    introspection_endpoint?: string;
+    revocation_endpoint?: string;
+    token_endpoint?: string;
+    userinfo_endpoint?: string;
+  };
+  registration_endpoint?: string;
+  request_object_signing_alg_values_supported?: Array<string>;
+  revocation_endpoint_auth_methods_supported?: Array<string>;
+  revocation_endpoint_auth_signing_alg_values_supported?: Array<string>;
+  revocation_endpoint?: string;
+  token_endpoint_auth_methods_supported?: Array<string>;
+  token_endpoint_auth_signing_alg_values_supported?: Array<string>;
+}
+
+/**
+ * A custom fetch(..) function that can be used when calling OIDC endpoints.
+ * Primarily used for intercepting responses or configuring an http proxy.
+ */
+export interface FetchFunction {
+  (input: string | URL, init?: FetchFunctionInit): Promise<Response>;
+}
+
+/**
+ * Init options for FetchFunction.
+ */
+export interface FetchFunctionInit extends RequestInit {
+  body?: string; // forces compatibility between node.fetch() and undici.fetch()
+}
+
+/**
+ * JsonWebKey set.
+ * Returned from the auth server's JWKS endpoint.
+ */
+interface JWKSet {
+  keys: Array<JsonWebKey>;
+}
+
+/**
+ * Fetches an auth server's metadata via its well-known discovery URL.
+ *
+ * @see https://datatracker.ietf.org/doc/html/rfc8414#section-3
+ */
+export async function fetchServerMetadata(authServerUrl: string, fetchFn?: FetchFunction) {
+  const discoveryUrl = authServerUrl + '/.well-known/openid-configuration';
+  log.info(`Fetching OIDC server metadata from [${discoveryUrl}]`);
+
+  // prettier-ignore
+  const discoveryResponse = fetchFn
+    ? await fetchFn(discoveryUrl)
+    : await fetch(discoveryUrl);
+
+  if (discoveryResponse.status !== 200) {
+    throw new Error('Error fetching server metadata: non-200 status');
+  }
+
+  const serverMetadata = (await discoveryResponse.json()) as ServerMetadata;
+  validateServerMetadata(serverMetadata);
+  log.silly(`Server metadata response: [${JSON.stringify(serverMetadata)}]`);
```

Comment:
Use `%j` instead of `JSON.stringify`.

https://github.com/winstonjs/winston?tab=readme-ov-file#string-interpolation
https://nodejs.org/dist/latest/docs/api/util.html#util_util_format_format_args

```suggestion
  log.silly('Server metadata response: [%j]', serverMetadata);
```

Reply:
Fixed.. please re-review